HowTo write a view (#841)
seqan/seqan3: doc/howto/write_a_view/index.md

# How to write a view {#howto_write_a_view}

[TOC]

This HowTo documents how to write a view using the standard library and some helpers from SeqAn3.

\tutorial_head{Difficult, 120 min, \ref tutorial_concepts\,\ref tutorial_ranges, }

\note
Some of the links from this HowTo only resolve in the developer documentation because

> they referred to entities from the seqan3::detail namespace.

Reviewer: ```suggestion
they refer to entities from the seqan3::detail namespace.
```

We recommend you open this tutorial from [the developer documentation](http://docs.seqan.de/seqan/3.0.0-master-dev/).

# Motivation

We have introduced "views" in \ref tutorial_ranges.
You can do many things with the views provided by the standard library and those shipped with SeqAn, but in
certain situations you will want to define your own view.
This page will teach you the basics of defining your own view.

# What makes a view?

A view is a type of `std::ranges::Range` that also models `std::ranges::View`.
The additional requirements of `std::ranges::View` can be vaguely summarised as "not holding any own data" or at least
not holding data that is relative in size to the number of elements in the view (e.g. a vector cannot be a view,

> because it's size in memory depends on the number of elements it represents).

Reviewer: ```suggestion
because its size in memory depends on the number of elements it represents).
```

> A simple example of aview is `std::ranges::subrange`. It can be constructed from a pair of iterators or more precisely

Reviewer: ```suggestion
A simple example of a view is `std::ranges::subrange`. It can be constructed from a pair of iterators or more precisely
```

That's why for every view that adapts an existing view we have an additional *adaptor object*, usually available
in a `view::` sub-namespace:

```cpp
std::vector vec{1, 2, 3, 4, 5};

auto l = [] (int i) { return i + 1; };

auto v = vec | std::view::transform(l);
```

This adaptor object (`std::view::transform`) provides the pipe operator and returns an object of the actual view type
(`std::ranges::transform_view`).
The pipe operator allows us to chain multiple adaptors similar to the unix command line.
We will discuss the details of these adaptor objects in the following sections.

# Custom range adaptor objects

Read [section 24.7 and 24.7.1 of the C++ standard](http://eel.is/c++draft/range.adaptors).

The wording of the standard needs some getting used to, but some important notes for us are:
  1. You can pipe a viewable range into series of *adaptor closure objects* and will get back a view (this is also
  what we did above):
```cpp
std::vector vec{1, 2, 3, 4, 5};
auto v = vec | std::view::transform([] (int i) { return i + 1; })
             | std::view::filter([] (int i) { return i % 2 == 0; });
// v is a view, you can iterate over it!
```

>   2. You can create a new *adaptor closure object* from an adaptor object that requires parameters, by providing those:

Reviewer: ```suggestion
  2. You can create a new *adaptor closure object* from an adaptor object that requires parameters by providing those:
```

// a is an adaptor and can be used as such:
auto v = vec | a;
```

| Terminology                    | Description                                                                                          | Example                                                                       |
|--------------------------------|------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------|
| "view"                         | A type that models std::ranges::View.                                                                | `std::ranges::filter_view<std::subrange<int const *, int const *>>`           |
| "view adaptor object"          | An object that returns a view via `operator()` and `operator│`; possibly with additional parameters  | `std::view::filter` and <br> `(std::view::filter([] (int) { return true; }))` |
| "view adaptor closure object"  | An object that returns a view via `operator()` and `operator│`; no additional paramaters             | `(std::view::filter([] (int) { return true; }))`                              |

> In  many cases where you are planning on creating "a new view", it will be sufficient to use the previously

Reviewer: ```suggestion
In many cases where you are planning on creating "a new view", it will be sufficient to use the previously
```

Some things to keep in mind for the implementation:
  * When defining a type template (like above), member types are not inherited implicitly (member functions are).
  * Another reason "just inheriting" is not sufficient, is that some inherited functions return objects or references
    to the base type, not your type (which is required by the `std::ForwardIterator` concept).
  * If you choose to wrap the underlying iterator instead of inheriting, you will need to define and "forward"
    a few more member functions, but it's good practice and will help you understand the concept.
  * The `static_assert` will just return `true` or `false`, to get more detailed information on why your type does
    not (yet) model the concept, implement a constrained function template and pass an object of your type to that
    template.
  * Your iterator needs to be comparable to the sentinel of the underlying range, you can access that type via
    `std::ranges::sentinel_t<urng_t>`.

<small>¹ In some situations it might be better to wrap the underlying iterator instead of inheriting it, i.e.
save a copy of the underlying iterator as a data member of your iterator.

> Reason could be that you *don't* want to inherit some members or want to prevent implicit convertibility.</small>

Reviewer: ```suggestion
A reason could be that you *don't* want to inherit some members or want to prevent implicit convertibility.</small>
```

```cpp
    using reference             = typename std::iterator_traits<base_t>::reference;

```
\snippet doc/howto/write_a_view/solution_iterator.cpp solution1b
\snippet doc/howto/write_a_view/solution_iterator.cpp end

The program prints "G A T T A C A ".
\endsolution

\assignment{Exercise 4: A transforming iterator}

In the previous assigment you have created a working – but pointless – iterator.
It does not do anything differently from the original.

Your task now is to implement the "complementing" behaviour, i.e.
  * your iterator should only work on ranges of `seqan3::NucleotideAlphabet`
  * when accessing an element through the iterator it should not return the element from the underlying range but
    instead its complement

Think about which operator is responsible for returning the element and be careful with the return type
of that operator as your transformation might make a change necessary.
\endassignment

\solution

1. The restriction on the alphabet type is done via a `static_assert`:
\snippet doc/howto/write_a_view/solution_iterator.cpp static_assert
You could have done this via an additional template constraint, too, but `static_assert` gives you the opportunity

> to give a readable message in case of error.¹

Reviewer: ```suggestion
to give a readable message in case of an error.¹
```

# How to write a view {#howto_write_a_view}

[TOC]

This HowTo documents how to write a view using the standard library and some helpers from SeqAn3.

> \tutorial_head{Difficult, 120 min, \ref tutorial_concepts\,\ref tutorial_ranges, }

Reviewer: ```suggestion
\tutorial_head{Hard, 120 min, \ref tutorial_concepts\,\ref tutorial_ranges, }
```
The difficulty is difficult 🤓

Author: But "hard" is more colloquial,

Reviewer: High then?

Author: Then we need to change the others to "low" and "medium". I don't have a problem with "Difficulty: difficult" though.

\note

> Some of the links from this HowTo only resolve in the developer documentation, because

Reviewer: ```suggestion
Some of the links from this HowTo only resolve in the developer documentation because
```

they referred to entities from the seqan3::detail namespace.

> We recommend you open this tutorial from http://docs.seqan.de/seqan/3.0.0-master-dev/

Reviewer: ```suggestion
We recommend you to open this tutorial from [the developer documentation](http://docs.seqan.de/seqan/3.0.0-master-dev/)
```

Author: the second half of the suggestion yes, the first no :)
This is more polite.

> \endnote

Reviewer: remove

Author: why?

Reviewer: `\endnote` doesn't exist.
You should also see a warning about it.

Author: Hm, that's confusing, why do the other all have \end but not \note? Or is note not a macro defined by us?

Reviewer: @smehringer is responsible for this heresy

Reviewer 2: ![image](https://user-images.githubusercontent.com/7955878/55020828-f526fc00-4ff7-11e9-8e6e-4dd1526696a8.png)

no I am not

Reviewer: Avenja means to say that this is a doxygen macro.

# Motivation

We have introduced "views" in \ref tutorial_ranges.
You can do many things with the views provided by the standard library and those shipped with SeqAn, but in
certain situations you will want to define your own view.
This page will teach you the basics of defining your own view.

# What makes a view?

A view is a type of `std::ranges::Range` that also models `std::ranges::View`.
The additional requirements of `std::ranges::View` can be vaguely summarised as "not holding any own data" or at least
not holding data that is relative in size to the number of elements in the view (e.g. a vector cannot be a view,
because it's size in memory depends on the number of elements it represents).

A simple view is `std::ranges::subrange`. It can be constructed from a pair of iterators or more precisely an

> *iterator* and a *sentinel*, i.e. the "iterator" representing the end need not be of the same type as the iterator

Author: I can fine-tune the sentence, but we should really introduce these things in the general ranges tutorial. I'll make a card.

representing the beginning – as long as they are comparable.
This view then holds exactly the iterator-sentinel-pair as its state and nothing else.

> But this type does not yet facilitate any "composing-behaviour" that you have seen in \ref tutorial_ranges tutorial,

Author: ok

`std::ranges::subrange` is simply a type that can be constructed from an iterator-sentinel-pair, you cannot "pipe"
anything into it.

> Most views are adaptors on other views, e.g. `std::ranges::transform_view` wraps an existing view and applies

Author: well not really. Strictly speaking they only wrap other views.

an element-wise transformation on-demand.
You can directly construct `std::ranges::transform_view` from another view, e.g.

> ```cpp

Reviewer: Snippet

Author: these snippets don't work yet, because
  1. we don't have transform_view in our std-module
  2. it's not trivial to alias, because like subrange, the deduction won't work

This would make the snippet really terrible to read. I suggest we open a card for finding a better solution to this later.

Reviewer: Mhm, well ... I'd kinda expect it to be copy and pasteable. :/

std::vector vec{1, 2, 3, 4, 5};

auto l = [] (int i) { return i + 1; };

std::ranges::transform_view v{std::ranges::subrange{vec.begin(), vec.end()}, l};
```

But this syntax gets difficult to read quickly which is why for every view that adapts an existing view
we have an additional *adaptor object*, usually available in a `view::` sub-namespace:

> ```cpp

Reviewer: Snuppet

Author: see above

std::vector vec{1, 2, 3, 4, 5};

auto l = [] (int i) { return i + 1; };

auto v = std::ranges::subrange{vec.begin(), vec.end()} | std::view::transform(l);
```

This adaptor object (`std::view::transform`) provides the pipe operator and returns an object of the actual view type
(`std::ranges::transform_view`).
The pipe operator allows us to chain multiple adaptors similar to the unix command line.
We will discuss the details of these adaptor objects in the following sections.

# Custom range adaptor objects

Read [section 24.7 and 24.7.1 of the C++ standard](http://eel.is/c++draft/range.adaptors).

Even if you didn't understand all details of the standard wording, you may have learned that you can combine
existing adaptor objects into new adaptor objects with the pipe operator (but without actually piping a range
into the whole thing, i.e. you just get a new adaptor object).

> And that for those adaptor objects that take multiple parameters (not the "range adaptor *closure* objects"), you

Reviewer: Either

```suggestion
And that for those adaptor objects that take multiple parameters (not the \"range adaptor <i>closure</i> objects\"), you
```

Reviewer: Or

```suggestion
And that for those adaptor objects that take multiple parameters (not the &quot;range adaptor *closure* objects&quot;), you
```

Author: I hate html :cry

> And that for those adaptor objects that take multiple parameters (not the "range adaptor *closure* objects"), you

Author: I have completely reformulated this paragrph, please re-review after the next push

can create a new adaptor object by fixing the parameters other than the view parameters.

> In  many cases where you are planning on creating "a new view" it will be sufficient to use the previously

Reviewer: ```suggestion
In  many cases where you are planning on creating "a new view", it will be sufficient to use the previously
```

mentioned techniques to just create "a new adaptor object" and not having to specify the actual view type yourself.

Let's look at some examples!

\assignment{Exercise 1: Your first custom adaptor object}

> You have learned in the alphabet module that ranges of alphabets are not implicitly

Reviewer: ```suggestion
In the alphabet module you have learned that ranges of alphabets are not implicitly
```

convertible to `char` so you **cannot** print a `std::vector<seqan3::dna5>` via `std::cout`¹.
You also know that you can call `seqan3::to_char` on every object that models `seqan3::Alphabet` which will convert it

> to `char` or a similar type.

Author: I did not want to divert to "underlying char type" here.

We want to do the following:

> ```cpp

Reviewer: Snoppet

#include <iostream>
#include <seqan3/alphabet/nucleotide/all.hpp>
#include <seqan3/std/ranges>

using namespace seqan3;

// your implementation goes here

int main()
{
    std::vector<dna5> vec{"ATTAGATTA"_dna5};
    // std::cout << vec[0] << '\n';                 // won't work

    auto v = vec | my_convert_to_char_view;

    std::cout << v[0] << '\n';                      // should be "A"
}
```

Define a *range adaptor object* that applies a concrete transformation (calling seqan3::to_char) on every element.

\hint
You need to need use `std::view::transform` and you need to set a fixed transformation function. `std::view::transform`
takes an object that models `std::RegularInvocable`, e.g. a lambda function with empty capture `[]`.
\endhint

<small>¹ You *can* print via `seqan3::debug_stream`, but let's ignore that for now. </small>
\endassignment

\solution

> ```cpp

Reviewer: Snappet

#include <iostream>
#include <seqan3/alphabet/nucleotide/all.hpp>
#include <seqan3/std/ranges>

using namespace seqan3;

constexpr auto my_convert_to_char_view = std::view::transform([] (auto const alph)
{
    return to_char(alph);
});

int main()
{
    std::vector<dna5> vec{"ATTAGATTA"_dna5};
    // std::cout << vec[0] << '\n';                 // won't work

    auto v = vec | my_convert_to_char_view;

    std::cout << v[0] << '\n';                      // prints "A"
}
```

You simply define your adaptor type as `auto` and make it behave like `std::view::transform`, except that you
"hard-code" the lambda function that is applied to each element.

> The object is marked as `constexpr`, because the adaptor object itself never changes, it only

Reviewer: ```suggestion
The object is marked as `constexpr` because the adaptor object itself never changes, it only
```

provides `operator()` and `operator|` that each return a specialisation of `std::ranges::transformation_view`.
\endsolution

\assignment{Exercise 2: Combining two existing adaptor objects}

> Study the `seqan3::NucleotideAlphabet`. It states that on all nucleotides you can call `seqan3::complement`

Reviewer: ```suggestion
Study the `seqan3::NucleotideAlphabet`. It states that you can call `seqan3::complement` on all nucleotides
```

which will give you <tt>'A'_dna5</tt> for <tt>'T'_dna5</tt> a.s.o. Think about how you can adapt the previous solution
to write a view that transforms ranges of nucleotides into their complement.

BUT, we are also interested in *reversing* the range which is possible with `std::view::reverse`:

> ```cpp

Reviewer: Sneppet

#include <seqan3/alphabet/nucleotide/all.hpp>
#include <seqan3/io/stream/debug_stream.hpp>
#include <seqan3/std/ranges>

using namespace seqan3;

// your implementation goes here

int main()
{

>     std::vector<dna5> vec{"ACCAGATTA"_dna5};

Reviewer: You could also make it `const`.
 (Once #839 is merged)

    debug_stream << vec << '\n';                    // will print "ACCAGATTA"

    auto v = vec | my_reverse_complement;

    debug_stream << v << '\n';                      // should print "TAATCTGGT"
}
```

Define a *range adaptor object* that presents of a view of the reverse complement of whatever you pipe into it.
\endassignment

\solution

> ```cpp

Reviewer: Snuppit

#include <seqan3/alphabet/nucleotide/all.hpp>
#include <seqan3/io/stream/debug_stream.hpp>
#include <seqan3/std/ranges>

using namespace seqan3;

auto my_reverse_complement = std::view::reverse | std::view::transform([] (auto const d)
{
    return complement(d);
});

int main()
{
    std::vector<dna5> vec{"ACCAGATTA"_dna5};
    debug_stream << vec << '\n';                    // will print "ACCAGATTA"

    auto v = vec | my_reverse_complement;

    debug_stream << v << '\n';                      // prints "TAATCTGGT"
}
```

> The adaptor consists of `std::view::reverse` combined with `std::view::transform`, this time the lambda just

Reviewer: ```suggestion
The adaptor consists of `std::view::reverse` combined with `std::view::transform`. This time the lambda just
```

\endsolution

# A full custom view implementation

Using existing adaptors only works to a certain degree, sometimes you will need to implement a full view.
As we have seen above, it is easy to implement a view that presents the complement of a nucleotide range just using
existing adaptors.
We will still use this as an example for a full implementation so we will create a non-generic transform view in
the next steps.

A full view implementation typically consists of the following components:

  1. The view class template that takes the underlying range as a template parameter.
     * The class should be derived from `std::ranges::view_interface` which will take care of a lot of boilerplate for
       us.
     * The actual functionality of the view is implemented in its iterator and sentinel.
  2. The type definition of the *adaptor object* which provides the operators.
  3. The *adaptor object* which is just an instance of the type defined in 2.

## The view class template

A good thing to start with is to think about the iterator and the sentinel of your view.
The iterator and/or its relation to the sentinel is the way we implement the behaviour that is specific to
our view.

> We will start with implementing the iterator seperately and later integrate it into the view.

Reviewer: ```suggestion
We will start with implementing the iterator separately and later integrate it into the view.
```

### Iterator

Since we know that in our current usecase we have exactly one element in our view for every element in the
underlying range, we don't need to change the relation between iterator and sentinel, i.e. "begin == end" on
our view iff "begin == end" on the underlying range.
This indicates that we can use the sentinel of the underlying range as-is.

\assignment{Exercise 3: Your first iterator}

> Have a peak at the \ref tutorial_concept tutorial again and study the `std::ForwardIterator` concept thoroughly.

Reviewer: ```suggestion
Have a peak at the \ref tutorial_concepts tutorial again and study the `std::ForwardIterator` concept thoroughly.
```

> Have a peak at the \ref tutorial_concept tutorial again and study the `std::ForwardIterator` concept thoroughly.

Author: No, I don't want to spend too much time on that as we will remove al of the std documentation as soon as cppreference has it anyway.

You will now have to implement your own forward iterator.

> \snippet doc/howto/write_a_view/solution_it.cpp start

Reviewer: ```suggestion
\snippet doc/howto/write_a_view/solution_iterator.cpp start
```

```cpp
    // YOUR IMPLEMENTATION GOES HERE (OPERATORS, MEMBER TYPES...)
```

> \snippet doc/howto/write_a_view/solution_it.cpp end

Reviewer: ```suggestion
\snippet doc/howto/write_a_view/solution_iterator.cpp end
```

> It should wrap the iterator of the underlying range or inherit from it (shown above); in the end, you should be able

Author: I have changed the text...

to iterate over the underlying range and print elements like you would with the original iterator, i.e. your iterator
shall behave exactly as the original in this regard (no transformation, yet).

Some things to keep in mind for the implementation:
  * When defining a type template (like above), member types are not inherited implicitly (member functions are).

>   * Another reason "just inheriting" is not sufficient, is that some inherited functions return objects or references

Reviewer: ```suggestion
  * Another reason "just inheriting" is not sufficient is that some inherited functions return objects or references
```

Author: hm, I am not sure about this one

Reviewer: mhm, I'd remove it.

Another reason [...] is that blablabla
Another reason [...], is that blablabla

But I guess people will get the message either way

    to the base type, not your type (which is required by the `std::ForwardIterator` concept).
  * If you choose to wrap the underlying iterator instead of inheriting, you will need to define and "forward"
    a few more member functions, but it's good practice and will help you understand the concept.
  * The `static_assert` will just return `true` or `false`, to get more detailed information on why your type does
    not (yet) model the concept, implement a constrained function template and pass an object of your type to that
    template.

>   * Your iterator needs to be comparable to the sentinel of the underlying range, you can access that type via

Author: hm, but its important to remember or not?

    `std::ranges::sentinel_t<urng_t>`.

\endassignment

\solution
\snippet doc/howto/write_a_view/solution_iterator.cpp start
\snippet doc/howto/write_a_view/solution_iterator.cpp solution1a
```cpp

>     using reference             = typename std::iterator_traits<base_t>::reference;

Author: Yeah, doxygen :rage2: 

Nothing I can do about it.

```
\snippet doc/howto/write_a_view/solution_iterator.cpp solution1b
\snippet doc/howto/write_a_view/solution_iterator.cpp end

The program prints "G A T T A C A ".
\endsolution

\assignment{Exercise 4: A transforming iterator}

> In the previous assigment you have created a working, but pointless iterator.

Reviewer: ```suggestion
In the previous assigment you have created a working but pointless iterator.
```

Author: Hm, yes, but then I rather use – or parens...

Reviewer: Sure

It does not do anything differently from the original.

Your task now is to implement the "complementing" behaviour, i.e.
  * your iterator should only work on ranges of `seqan3::NucleotideAlphabet`

>   * when accessing an element through the iterator it should not return the element from the underlying range, but

Reviewer: ```suggestion
  * when accessing an element through the iterator it should not return the element from the underlying range but
```

>     instead it's complement

Reviewer: ```suggestion
    instead its complement
```

> Think about which operator is responsible for returning the element and have special care with the return type

Reviewer: ```suggestion
Think about which operator is responsible for returning the element and take special care of the return type
```

Author: I meant "give a care" (*gib acht*) not "take care" (*kümmern*). But i will change it to something else altogether.

Reviewer: Ok.

You could have done this via an additional template constraint, too, but `static_assert` gives you the opportunity
to give a readable message in case of error.¹

2. The operator that needs to call the `seqan3::complement` function is `operator*`:
\snippet doc/howto/write_a_view/solution_iterator.cpp dereference

3. As previously noted, care needs to be taken with this function's return type:
\snippet doc/howto/write_a_view/solution_iterator.cpp reference

Here is the full solution:
\hint
\include doc/howto/write_a_view/solution_iterator.cpp

The program prints "C T A A T G T "

\endhint

<small>¹ This is only recommended when you do not want to allow a different specialisation of the template to
cover the excluded case.</small>

\endsolution

You now have a working iterator, although it still lacks the capabilities of `std::BidirectionalIterator`,
`std::RandomAccessIterator` and `std::ContiguousIterator`.
When designing views, you should always strive to preserve as much of the capabilities of the underlying range
as possible.

Which of the mentioned concepts do you think your iterator could be made to implement? Have a look at the respective

> documentation.

Reviewer: `std::ContiguousIterator` does not link /  not available yet

Author: I'll add it to std/ranges

\hint
It could be designed to be a `std::RandomAccessIterator` (and thus also `std::BidirectionalIterator`) when the
underlying range is, because jumping on your iterator/view can be done in
constant time iff it can be done in constant time on the underlying range (you just jump to the n-th element
of the underlying range and perform your transformation one that).

> However, it can never model `std::ContiguousIterator`, because that would imply that the elements are adjacent to each

Reviewer: ```suggestion
However, it can never model `std::ContiguousIterator` because that would imply that the elements are adjacent to each
```

other in memory (the elements of our view are created on demand and are not stored in memory).
\endhint

> If you have looked at the `std::RandomAccessIterator` you will have seen that it is quite a bit of work to implement

Reviewer: ```suggestion
If you have looked at the `std::RandomAccessIterator`, you will have seen that it is quite a bit of work to implement
```

> all the operators, many of whome just need to be overloaded to fix the return type.

Reviewer: ```suggestion
all the operators, many of whom just need to be overloaded to fix the return type.
```

> To make this a little bit easier SeqAn provides `seqan3::detail::inherited_iterator_base`, it fixes the issue with the

Reviewer: ```suggestion
To make this a little bit easier SeqAn provides `seqan3::detail::inherited_iterator_base` - it fixes the issue with the
```

return type via CRTP.

> A solution to the previous exercise look like this:

Reviewer: ```suggestion
A solution to the previous exercise looks like this:
```

\snippet doc/howto/write_a_view/solution_view.cpp iterator
\snippet doc/howto/write_a_view/solution_view.cpp main
\snippet doc/howto/write_a_view/solution_view.cpp end

### The view class

We now implement the view in several steps:

\snippet doc/howto/write_a_view/solution_view.cpp view_header

Like the iterator, the view is derived from a CRTP base class that takes care of defining many members

> for us, e.g. `.size()`, `.operator[]` and a few others.

Author: I am not sure what you mean exactly. This place is about range_interface not inherited_iterator_base. We will get cppreference documentation for range_iterface sometime soon I hope...

\snippet doc/howto/write_a_view/solution_view.cpp view_private

The only data member the class holds is a copy of the underlying range.
As you may have noted above, our class only takes underlying ranges that model
std::ranges::View.

> This might seem strange, after all we want to apply the view to a vector of which we know that it

Reviewer: ```suggestion
This might seem strange; after all we want to apply the view to a vector of which we know that it
```

is not a view, but we will clear this up later.

\snippet doc/howto/write_a_view/solution_view.cpp view_member_types

The only member types that we define here are the definitions of the iterators which are just the iterator
we have defined before.
Note that we have `const_iterator` in addition to `iterator` which const-qualified member functions return,

> because in a const-context the `urange` data member will be const we cannot return the mutable `iterator` from it.

Reviewer: ```suggestion
because in a const-context the `urange` data member will be const so we cannot return the mutable `iterator` from it.
```

\snippet doc/howto/write_a_view/solution_view.cpp view_end

The picture for `end()` is similiar except that for our range the sentinel type (the return type of `end()`) is
the same as of the underlying range, we just pass it through.

For many more complex views you will have to define the sentinel type yourself or derive it from the underlying
type in a similar manner to how we derived the iterator type.
Often you can use `std::ranges::default_sentinel_t` as the type for your sentinel and implement the
"end-condition" in the iterator's equality comparison operator against that type.

\snippet doc/howto/write_a_view/solution_view.cpp view_constructors

We have two constructors, one that takes an the underlying type by copy and moves it into the data member
(remember that since it is a view, it will not be expensive to copy – if it is copied).

The second constructor is more interesting, it takes a `std::ranges::ViewableRange` which is defined as
being either a `std::ranges::View` or a reference to `std::ranges::Range` that is not a view.
Since we have a constructor for `std::ranges::View` already, this one explicitly handles the second case

> and goes through `std::view::all` which wraps the reference in a thin view-layer.

Author: ![image](https://user-images.githubusercontent.com/2996863/55033039-85723a80-5012-11e9-948d-df6c760c136f.png)

> Storing only a view member guarantess that our type itself is also cheap t copy among other things.

Reviewer: ```suggestion
Storing only a view member guarantess that our type itself is also cheap to copy among other things.
```

Note that both of these constructors seem like generic functions, but they just handle the underlying type or a
type that turns into the underlying when wrapped in `std::view::all`.

\snippet doc/howto/write_a_view/solution_view.cpp view_deduction_guide

> To easily use the second constructor we need to provide type deduction guide.

Reviewer: ```suggestion
To easily use the second constructor we need to provide a type deduction guide.
```

```cpp
std::vector vec{1, 2, 3, 4, 5};
auto a = std::view::transform([] (int i) { return i + 1; });
// a is an adaptor and can be used as such:
auto v = vec | a;
```

  4. You can create a new *adaptor closure object* from existing adaptor closure objects via `|` but without providing
  a range:
```cpp
std::vector vec{1, 2, 3, 4, 5};
auto a = std::view::transform([] (int i) { return i + 1; })
       | std::view::filter([] (int i) { return i % 2 == 0; });
// a is an adaptor and can be used as such:
auto v = vec | a;
```

| Terminology                    | Description                                                             | Example                                                                       |
|--------------------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------|
| "view"                         | A type that models std::ranges::View.                                   | `std::ranges::filter_view<std::subrange<int const *, int const *>>`           |

> | "view adaptor object"          | Creates a view; can be combined with other adaptors.                    | `std::view::filter` and <br> `(std::view::filter([] (int) { return true; }))` |

Reviewer 2: ```suggestion
| "view adaptor object"          | Creates a view; can be combined with other adaptors.                    | `seqan3::vew::reverse` `std::view::filter` and <br> `(std::view::filter([] (int) { return true; }))` |
```

> | "view adaptor closure object"  | A "view adaptor object" that takes no paramaters other than the range.  | `(std::view::filter([] (int) { return true; }))`                              |

Reviewer 2: ```suggestion
| "view adaptor closure object"  | A "view adaptor object" that takes no paramaters other than the range.  | `seqan3::vew::reverse`, `(std::view::filter([] (int) { return true; }))` but not `std::vew::filter`                              |
```

In many cases where you are planning on creating "a new view", it will be sufficient to use the previously
mentioned techniques to just create "a new adaptor object" and not having to specify the actual view type yourself.

Let's look at some examples!

\assignment{Exercise 1: Your first custom adaptor object}

> In the alphabet module you learned that ranges of alphabets are not implicitly

Reviewer 2: ```suggestion
In the alphabet module you've learned that ranges of alphabets are not implicitly
```

convertible to `char` so you **cannot** print a `std::vector<seqan3::dna5>` via `std::cout`¹.
You also know that you can call `seqan3::to_char` on every object that models `seqan3::Alphabet` which will convert it
to `char` or a similar type.

We want to do the following:

\snippet doc/howto/write_a_view/view_exercise1.cpp start
```cpp
// your implementation goes here
```
\snippet doc/howto/write_a_view/view_exercise1.cpp end

Define a *range adaptor object* using an existing adaptor which applies a concrete transformation
(calling seqan3::to_char) on every element.

\hint
You need to need use `std::view::transform` and you need to set a fixed transformation function. `std::view::transform`
takes an object that models `std::RegularInvocable`, e.g. a lambda function with empty capture `[]`.
\endhint

<small>¹ You *can* print via `seqan3::debug_stream`, but let's ignore that for now. </small>
\endassignment

\solution

\include doc/howto/write_a_view/view_exercise1.cpp

You simply define your adaptor type as `auto` and make it behave like `std::view::transform`, except that you

> "hard-code" the lambda function that is applied to each element.

Reviewer 2: Add: Since your adaptor object now only takes a range as the only parameter, it is an adaptor closure object.

The object is marked as `constexpr` because the adaptor object itself never changes, it only
provides `operator()` and `operator|` that each return a specialisation of `std::ranges::transformation_view`.
\endsolution

\assignment{Exercise 2: Combining two existing adaptor objects}

Study the `seqan3::NucleotideAlphabet`. It states that you can call `seqan3::complement` on all nucleotides
which will give you <tt>'A'_dna5</tt> for <tt>'T'_dna5</tt> a.s.o. Think about how you can adapt the previous solution
to write a view that transforms ranges of nucleotides into their complement.

BUT, we are also interested in *reversing* the range which is possible with `std::view::reverse`:

\snippet doc/howto/write_a_view/view_exercise2.cpp start
```cpp
// your implementation goes here
```
\snippet doc/howto/write_a_view/view_exercise2.cpp end

> Define a *range adaptor object* that presents of a view of the reverse complement of whatever you pipe into it.

Reviewer 2: o_O

Since we know that in our current usecase we have exactly one element in our view for every element in the
underlying range, we don't need to change the relation between iterator and sentinel, i.e. "begin == end" on
our view iff "begin == end" on the underlying range.
This indicates that we can use the sentinel of the underlying range as-is.

\assignment{Exercise 3: Your first iterator}

Have a peak at the \ref tutorial_concepts tutorial again and study the `std::ForwardIterator` concept thoroughly.
You will now have to implement your own forward iterator.

\snippet doc/howto/write_a_view/solution_iterator.cpp start

```cpp
    // YOUR IMPLEMENTATION GOES HERE (OPERATORS, MEMBER TYPES...)
```
\snippet doc/howto/write_a_view/solution_iterator.cpp end

In order to re-use functionality of the underlying range's iterator type you can inherit from it
(`std::ranges::iterator_t` returns the iterator type).¹ In the end, you should be able
to iterate over the underlying range and print elements like you would with the original iterator, i.e. your iterator
shall behave exactly as the original in this regard (no transformation, yet).

Some things to keep in mind for the implementation:

>   * When defining a type template (like above), member types are not inherited implicitly (member functions are).

Reviewer 2: ```suggestion
  * When defining a type template (like above), member **types** are not inherited implicitly (member functions are).
```

\snippet doc/howto/write_a_view/solution_iterator.cpp solution1a
```cpp
    using reference             = typename std::iterator_traits<base_t>::reference;

```
\snippet doc/howto/write_a_view/solution_iterator.cpp solution1b
\snippet doc/howto/write_a_view/solution_iterator.cpp end

The program prints "G A T T A C A ".
\endsolution

\assignment{Exercise 4: A transforming iterator}

In the previous assigment you have created a working – but pointless – iterator.
It does not do anything differently from the original.

Your task now is to implement the "complementing" behaviour, i.e.
  * your iterator should only work on ranges of `seqan3::NucleotideAlphabet`
  * when accessing an element through the iterator it should not return the element from the underlying range but
    instead its complement

Think about which operator is responsible for returning the element and be careful with the return type
of that operator as your transformation might make a change necessary.
\endassignment

\solution

1. The restriction on the alphabet type is done via a `static_assert`:
\snippet doc/howto/write_a_view/solution_iterator.cpp static_assert

> You could have done this via an additional template constraint, too, but `static_assert` gives you the opportunity

Reviewer 2: ```suggestion
You could have done this via an additional template constraint¹, too, but `static_assert` gives you the opportunity
```

> to give a readable message in case of an error.¹

Reviewer 2: ```suggestion
to give a readable message in case of an error.
```

As you may have noted above, our class only takes underlying ranges that model
std::ranges::View.
This might seem strange; after all we want to apply the view to a vector of which we know that it
is not a view, but we will clear this up later.

\snippet doc/howto/write_a_view/solution_view.cpp view_member_types

The only member types that we define here are the definitions of the iterators which are just the iterator
we have defined before.
Note that we have `const_iterator` in addition to `iterator` which const-qualified member functions return,
because in a const-context the `urange` data member will be const so we cannot return the mutable `iterator` from it.

Many ranges like the standard library containers also present the member types of the iterator, i.e. `value_type`,
`reference` a.s.o, but this is not required to model any of the range concepts.

\snippet doc/howto/write_a_view/solution_view.cpp view_begin

These functions are the same member functions you know from `std::vector`, they return objects of the previously
defined iterator types that are initialised with the begin iterator from the underlying range.

\snippet doc/howto/write_a_view/solution_view.cpp view_end

> The picture for `end()` is similiar except that for our range the sentinel type (the return type of `end()`) is

Reviewer 2: ```suggestion
The implementation for `end()` is similar except that for our range the sentinel type (the return type of `end()`) is
```

`view::my(FOO)` is identical to `my_view{FOO}`.

The second operator is declared as friend, because the left-hand-side of the `operator|` is generic, it's how
the range is handled in snippets like `auto v = vec | view::my`.

\note
This adaptor type does not yet provide the ability to combine multiple adaptors into a new adaptor, it only
handles ranges as left-hand-side input to `operator|`.

Our example adaptor type definition is rather simple, but for views/adaptors that take more parameters it gets quite
complicated quickly. Therefore SeqAn provides a convenience template that does all of this for you:
```cpp
//                                      your view type goes here ↓
using my_view_fn = seqan3::detail::generic_pipable_view_adaptor<my_view>;
```

### Adaptor object definition

The adaptor object is simply an instance of the previously defined type:

\snippet doc/howto/write_a_view/solution_view.cpp adaptor_object_definition

As noted above, we place this object in a `view::` sub-namespace by convention.
Since the object holds no state, we mark it as `constexpr` and since it's a global variable we also mark it as
`inline` to prevent linkage issues.

> Finally we can use our view in with pipes and combine it with others:

Reviewer 2: ```suggestion
Finally we can use our view with pipes and combine it with others:
```
